### ISSUE TITLE
Scripts are run unnecessarily depending on how turbo is configured

### ISSUE THREAD

**Issue by dcherman** [vercel/turborepo#937]
### What version of Turborepo are you using?

1.1.9

### What package manager are you using / does the bug impact?

Yarn v1

### What operating system are you using?

Mac

### Describe the Bug

When running a command like `turbo run test`, I would expect it to build the graph of tasks to run based on packages that actually define a `test` task.  Instead, it seems like it assumes that all packages contain that task which results in needlessly running all tasks that would be dependencies.

In the case of the reproducer below, only the `bar` package contains a `test` script with a dependency on `foo`.  The `build` task in baz was executed despite the output of it not being required for executing the requested tasks.

### Expected Behavior

Only packages that actually contain the tasks to run should be considered when building the task graph.

### To Reproduce

1. Clone https://github.com/dcherman/turborepo-bad-script-reproducer
2. `yarn install`
3. `yarn reproduce`

Since only `bar` has a `test` script, I expected to see the `build` task from `foo` and `bar` run followed by the `test` task from `bar`.  The `build` task of `baz` runs despite it having no `test` task and it not being a dependency of either of the other packages.

Here's an image of the graph that's generated from running `turbo run test --graph`

![image](https://user-images.githubusercontent.com/976608/160013487-6798132d-b8a6-47f3-bd97-17696a6a4a6b.png)

Note that the both the `foo` and `baz` tasks contain a node for the `test` task despite not actually defining a `test` task in package.json.

**Comment by gsoltis** [comment]
I can see the issue here, and I think I'm maybe 99% convinced the behavior should change. I am a little worried about anyone using "synthetic" tasks to trigger upstream packages to do something, but I think we have other ways of handling that (`--include-dependencies` / `--filter=<pkg>...` if it is desired behavior.

**Comment by jaredpalmer** [comment]
See https://turborepo.org/docs/features/filtering and the new `--filter` flag in `turbo@1.2` which should correct this. Sorry about that!

**Comment by michaelkplai** [comment]
I think this discussion may also be related #1106.

I also recently ran into this issue. The section of the documentation that mentions that missing tasks are gracefully ignored led me to assume missing task dependencies would also be ignored. See https://turborepo.org/docs/core-concepts/pipelines#tasks-that-are-in-the-pipeline-but-not-in-some-packagejson.

Since ignoring missing task dependencies would be a breaking change it may make sense to introduce this as a new cli flag or config option rather than overwrite the default behavior.

**CLI Option** `turbo run build --ignore-missing-task-deps` 

**turbo.json Option** `”ignoreMissingTaskDeps”: true`.

I’d be open to a more terse option name. However, `--if-present` could be a little confusing since the npm implementation is more about suppressing errors for missing scripts and has no relation to dependencies. See https://docs.npmjs.com/cli/v8/commands/npm-run-script#if-present.

I’d be happy to look into contributing a PR if that helps move this issue along.

**Comment by finn-orsini** [comment]
Joining the discussion here from a [similar issue that was marked as duplicate](https://github.com/vercel/turborepo/issues/1135). 

If I understand correctly, the current behavior can be described as: 

- Turbo will attempt to run tasks that do not exist, and "gracefully ignore" them.
- Turbo will attempt to run dependencies of tasks that do not exist.

and these two behaviors repeat themselves all the way down the pipeline dependency graph until no more potential dependencies are found.

This issue appears to be discussing the second bullet point, i.e. `should Turbo try to run dependencies of tasks that do not exist?` 

Although similar, #1135 is more related to the first bullet point, i.e. what does "gracefully ignoring" a task mean with respect to the `--dry-run` output? Currently the missing tasks are filtered from the overall task list if they don't exist, but they still appear as `dependencies` of existing tasks. 

I'm not completely convinced these are the same issue - but maybe there are some implementation details that would result in the decision here addressing both?

**Comment by gsoltis** [comment]
@finn-orsini you're right, they are technically distinct but related. 

I think we're going to move forward with a PR to prune packages that don't have the requested task, which will have the result of causing the output from `--dry-run` match what is actually executed. 

However, given that the above is technically a breaking change (workaround: define a no-op task), it may take a little while to land. In the meantime, I'll put up a PR to include the non-existent tasks in the `--dry-run` output, so that whichever way the above change goes, at least `--dry-run` will be more informative.

**Comment by zanona** [comment]
Would it be possible to have the expected behaviour under a flag i.e:`--if-present` and in the next major default to it instead?

**Comment by TxHawks** [comment]
I'd really like to see an intermediary step too. I was actually forced to rename some of my pipline tasks because of this behavior, which is far from ideal

**Comment by gsoltis** [comment]
I will look into putting this behind a flag.

**Comment by ObliviousHarmony** [comment]
> it sounds like your test task should explicitly "dependsOn": ["build", "^build"] if building dependencies are required for tests to run

I thought about this too @mantljosh, but, it falls apart with transitive dependencies. Even if my `"test"` task builds the direct dependencies of my package, when one of those _also_ has no `"build"` but has dependencies that do, we reach the same conclusion. In my case, there are packages pulled into the main app which are not built themselves but have JS dependencies that are.

**Comment by gsoltis** [comment]
Discussed a bit w/ @jaredpalmer and he pointed out that it's reasonable for pure-js packages to not have a build step, but still require build steps from their dependencies. Given that's a scenario that we want to support, I don't think we can proceed with this change as a default. 

Perhaps we could augment the `--filter` syntax to include matching on having a particular task? So you could run `turbo run test --filter=#has-task` or something like that. Or perhaps this could be configured per-task in `turbo.json`?

**Comment by ObliviousHarmony** [comment]
@gsoltis Augmenting the `--filter` syntax likely wouldn't be sufficient since you would want to be able to run `turbo run test --filter={package}` and have it exhibit the desired behavior too. What do you think of my token proposal above?

**Comment by dcherman** [comment]
@ObliviousHarmony It seems like surprising behavior that Turborepo generates the task pipeline as if every project contains a `build` task even if one or more project don't actually contain one.

Although this is absolutely a breaking change, you could accomplish this by adding a `build` task to your project that does nothing but `exit 0`.  At that point, your project actually has a built tasks and it's no longer surprising that it'd be included in the pipeline.

FWIW, I think the proposal to add `--if-present` to the current version with a default to `false` and potentially change it to default to `true` in the next major version is a great idea.

**Comment by dcherman** [comment]
@ObliviousHarmony The problem isn't solely that the dependencies don't have the task to run, so I'm not sure that your token proposal would actually solve the problem described in the original bug.  To fix that one, we need to filter the initial set of projects down to ones that actually have the task being run.  I see your point about transitive dependencies though and wonder if we'd end up needing a combination of both solutions.

FWIW, that `if-present` flag could be supported in config pretty easily if that's desirable - just move the flag to an `ifPresent` boolean property in the config file.  This could follow the same precedence rules as other options so that environment variables, config, and flags are all supported.

**Comment by ObliviousHarmony** [comment]
> so I'm not sure that your token proposal would actually solve the problem described in the original bug

Ah @dcherman, I came here from the open pull request and so I misunderstood the original bug it seems. No, my proposal wouldn't solve that.

I 100% agree that running the `"build"` task on projects that don't have a `"test"` task doesn't make any sense. There is still a case for running a `"build"` task on a project with no script but that seems like the edge case. Either a configuration option or the filter augmentation proposed by @gsoltis would work just fine. One benefit of the filtering approach is that it creates a base for the hash syntax to add modifiers to filtering.

**Comment by gsoltis** [comment]
One concern I have with using the configuration approach is the ability to override it. Currently, tasks in `turbo.json` map 1:1 to the name of the script being run, so it is not feasible to, for instance, have two copies of "test", with and without the option to include targets that don't have a "test" script. This means that regardless of whether or not we augment the config file, we _must_ have a flag to control behavior per-execution.

**Comment by mantljosh** [comment]
This is an annoying problem because like other's have mentioned, both approaches have benefits and downsides.

I think there may be a safe incremental step to take, where we only do the "ignoring" based off the top level of the execution graph. What I mean is that if I run `turbo run test`, it should immediately filter out any packages that do not have a `test` script and ignore them (and their dependencies) entirely.

**Comment by georeith** [comment]
This would break my use case:

I have tests that run from source, however I want the cache of those tests to miss if one of their dependencies changes regardless of whether that package itself has tests.

However I would say that `^tests` is not necessarily the right thing here either as I don't care if my topological dependencies test configuration changes.

Really what I want to express is that I want to use the input hash of `build` but I don't actually want to run `build`. There's no way of currently doing this I believe?

**Comment by dcherman** [comment]
Until this is addressed, here's how I've been working around this:

This script generates the set of filter flags needed to run any given task

```bash
targetScript="$1"
mapfile -t < <(find packages -name "package.json" | grep -Ev "(node_modules|bower_components)")
jq -r --arg targetScript "$targetScript" '[inputs] | map(select(.scripts[$targetScript] != null)) | map("--filter=" + .name) | join(" ")' "${MAPFILE[@]}"
```

Which you can then pass to CLI using expansion:

```bash
turbo run build ${FILTER_FLAGS[*]}
```

It just supports a single script since that's our use case, but modifying it for multi task support should be pretty easy if needed.

**Comment by tom-sherman** [comment]
> Discussed a bit w/ @jaredpalmer and he pointed out that it's reasonable for pure-js packages to not have a build step, but still require build steps from their dependencies.

@gsoltis In this scenario wouldn't it be reasonable to expect that you have to explicitly define `dependsOn: [^build]`?

**Comment by tom-sherman** [comment]
I have a slightly more minimal reproduction: https://github.com/tom-sherman/turborepo-dependency-bug

In this repo i have a graph as follows

```mermaid
graph TD
foo#test --> foo#build --> __root__
bar#build --> __root__
```

My pipeline configuration looks like this:

```json
{
  "pipeline": {
    "build": {
      "dependsOn": ["^build"],
      "outputs": ["dist/**", ".next/**"]
    },
    "test": {
      "dependsOn": ["build"],
      "outputs": []
    }
  }
}
```

Therefore when I `turbo run test` I expect `foo#build` and `foo#test` to run, however `bar#build` also runs unexpectedly.

`bar#build` is not an ancestor of any test task so I don't expect it to run.

**Comment by gsoltis** [comment]
@tom-sherman `turbo run test` is attempting to run `foo#test` and `bar#test`. Since `test` depends on `build`, `foo#build` and `bar#build` are considered necessary. In this case, it's obvious we don't actually want `bar#build` because we're running tests, and `bar` does not have any tests. 

The story is different with the `build` example. Currently, a package that has no build step of its own, but requires its dependencies to be built, looks very similar to the above case with `test`. I.e. filtering out the non-existent task first, before resolving dependencies, would result in that package's dependencies not being built, even with a dependency on `^build`. 

So the question is how we want to handle these two scenarios. In my opinion, as a default, doing extra work is probably preferable to not doing enough work, especially if the extra work is easily cache-able. So I think the current behavior should remain _as a default_. But, we can absolutely look at ways of providing configuration hooks (cli flags, env vars, `turbo.json` keys, etc.) to allow users to specify when they want to opt in to the desired behavior in your test example. 

As an additional note, there are two flavors of skipping undefined tasks, assuming the behavior has been opted-in to: 
 1. always skip them. Anytime you encounter an undefined task, consider it "done" and stop the dependency graph traversal
 2. only filter the tasks that have been specified and are not in the entrypoint packages. In this case, If you run `turbo run test`, and a package is missing a `test` script, it and its dependencies won't get run. But, if a dependency of a package with a `test` script is missing a `build` script, we would _not_ stop the traversal, and instead continue with building its dependencies.
 
I think the latter is more likely to work as expected. Tasks that really want to stop traversal can already opt in to that behavior on a per-package-task basis by defining a package-task with an empty `dependsOn`.

**Comment by tom-sherman** [comment]
@gsoltis I don't understand the build example, can you please create a minimal reproduction? Really struggling to get my head around how it's the same issue as what I've described and also matching my mental model up with what you're describing

**Comment by gsoltis** [comment]
I'll see if I can get a repo together, that's probably a good idea for this discussion.

But in general, `turbo` works by taking the user input for which tasks to run (`build`, `test`, etc.) and which packages to run them in ("entry points"), and then filling in the graph for what needs to be done to run those tasks in the entry point packages. 

The essence of this discussion is what to do when you encounter a node in that graph that doesn't define the task we're looking for. Do you stop traversing and consider it done? Or do you pretend that the script exists and continue traversing to what would be its dependencies if it did exist. So a package missing a `test` script and a task missing a `build` script face the same decision: do we run the dependencies, even though the script is missing. The trouble comes from the difference in expectations for the two cases. For the `build` case, I think consensus is "yes, run the dependencies", and for `test`, it looks like we would prefer not to run the dependencies, as there are no tests to run. But without `turbo` knowing the difference between `test` and `build`, it has no way to distinguish the two cases.

Some other distinctions that we can make for this behavior are "is this package one of the entry point packages" and "is this task one of the requested tasks" vs packages and tasks that are transitively included as dependencies.

**Comment by gsoltis** [comment]
@tom-sherman  Sorry for the delay, but I have put together a sample that I believe demonstrates the issue: https://github.com/gsoltis/sample-monorepo

There are two top-level applications, only one has `test` script. There is also a library that has no `build` script, but depends on another library that does have a `build` script.

**Comment by arturenault** [comment]
Any updates here? This breaks my use case too.

**Comment by mantljosh** [comment]
While I think there are some difficulties that have been raised (i.e. a plain-js package that still needs its dependencies to be built even if it has no build script), I think there is an improvement that can be safely made where if I run `turbo run test` then it will:

1. First filter the list of packages to only ones that have a `test` script
2. Then run the build in the same way that is already run today (with potentially unnecessary steps being run)

This would at least let us avoid the case where `turbo run test` ends up building the world, even for things that will never be used by the `test` scripts.

**Comment by gsoltis** [comment]
@arturenault can you describe what specifically is broken? Is `turbo` doing extra work, or is some task being run incorrectly?

My understanding of the current state of the world is that in the worst case, `turbo` does unnecessary work, and the difficulty is in distinguishing what exactly is unnecessary. As you can see in the thread, there are a variety of opinions on what the expected default behavior is, so we are currently defaulting to possibly doing more, rather than possibly doing less than expected.

**Comment by gsoltis** [comment]
@mantljosh are you proposing to special-case `test`? Or somehow mark it in `turbo.json` to opt in to that behavior? We need some way to distinguish it from the `build` case.

FWIW, I think a marker of some sort in `turbo.json` might be the way to solve this, and we could even default to setting that marker for `test` in `create-turbo`. I'm not sure I would support special-casing specific task names inside `turbo` though.

**Comment by arturenault** [comment]
@gsoltis My case did cause actual issues. We use serverless framework for some of our projects which packages things differently for different environments. So those projects have both a `build` and a `build:${ENV} ` script. Both depend only on `^build`. 

We have an `integration-tests` project which has no lifecycle scripts, but depends on `api`, which is a serverless framework project.

So when I run, e.g., `turbo run deploy:test` (` depends on `build:test`):
- `api:build:test` gets run as a dependency of `api:deploy:test`.
- `api:build` gets run as a dependency of the (nonexistent) `integration-tests:deploy:test`.

This causes two builds to (potentially) happen at the same time, one for the test environment and one for the (default) dev environment. This causes build failures because both builds try to write to the same files.

I suspect it could cause the build to deploy to the wrong environment, but I'm not sure if turbo reads files directly from the filesystem or if it checks the dependency tree to fetch outputs from the right dependent task.

For now I've been able to work around this by excluding `integration-tests` from all my deployment scripts. But it's definitely possible that my setup isn't ideal! We're migrating from lerna and I've just been adapting our existing commands.

**Comment by gsoltis** [comment]
@arturenault Making sure I understand:

`deploy:${ENV}` depends on `build:${ENV}`, that's why `api:deploy:test` causes `api:build:test` to run.

`integration-tests:deploy:test` doesn't exist, but depends on `build:test`, which depends on `^build`, so `api:build` is also triggered. That's where the environment specifier is getting lost?

It does seem like there's the potential for this setup to break if you have two top-level applications and one depends on the other. For instance, if a hypothetical `web` depends on `api`, trying to deploy both `web:deploy:test` and `api:deploy:test` this way will result in the same behavior: building both `api:build:test` and `api:build`.

In terms of what you're trying to accomplish right now, when you run `turbo run deploy:test`, are you trying to deploy more than one package? While this gets sorted out, perhaps there's a way to use `--filter` to get at the right set of packages? 

This also does highlight one current sharp edge: package-task combos that read or write to the same location need to have a dependency defined between them, or else they could get scheduled concurrently. It's something that's on our roadmap to detect automatically. I'm not sure that would fully solve your specific case, but it would at least fail in a more consistent state.

**Comment by mantljosh** [comment]
> @mantljosh are you proposing to special-case test? Or somehow mark it in turbo.json to opt in to that behavior? We need some way to distinguish it from the build case.

@gsoltis no there would be no special casing here - just when running `turbo run some-script` the first step should be to determine which packages even have a `some-script`, and only compute the tasks to run on those.

**Comment by gsoltis** [comment]
@mantljosh got it. We unfortunately can't make that the default behavior because some packages still need their dependencies to be built before they can consider themselves to have been built, even if they have no `build` script.

**Comment by georeith** [comment]
> Although this is absolutely a breaking change, you could accomplish this by adding a build task to your project that does nothing but exit 0. At that point, your project actually has a build task and it's no longer surprising that it'd be included in the pipeline.

This to me seems like the most sensible answer, pipelines would work how the docs explain them (IMO much more intuitively) and edge cases can explicitly opt-in.

Sadly there is not anyway to opt-out of this currently and this is a pretty critical issue when say you have 30 packages and want to run a script in 2 of them that depends on those packages being built.

I think most of us would rather not wait for a breaking change given the above, even if I feel like the breaking change is the more pure solution. So would like to +1 adding a configuration option to make turborepo filter packages by the existence of said script first.

**Comment by LionC** [comment]
Just to add another resulting problem to this (talking to @anthonyshew in Discord about this right now):

This will also lead to turbo reporting circular dependencies and thus not executing anything, even though with the actual tasks and dependencies given, there is no circular dependency and it would be executable just fine.

For an example, see this reproduction repository: https://github.com/LionC/turbo-circular-dependency-example

Here, with all tasks that are present in the workspaces, the graph *should* look like this

![image](https://user-images.githubusercontent.com/3419961/211298699-55deea4e-43b7-43a2-b41e-ae3719f10e73.png)

But instead it reports a circular dependency

```
 ERROR  run failed: error preparing engine: Invalid task dependency graph:
cyclic dependency detected:
	schema#build,___ROOT___#codegen,___ROOT___#build
```

**Comment by dobesv** [comment]
I think this issue is causing us a lot of unnecessary rebuilds.  Sure you can say the jobs are cacheable but keep in mind that they won't be cached if they depend on tasks that are being "virtually" re-run, even though those tasks don't exist.  So this problem adds a lot of a imaginary dependencies and causes a lot of unexpected rebuilds.

Some use cases that I have which are relevant here:

I have a build script that needs another tool to be built in order to run - if the script is defined for that package - and which depends on some tasks from its dependencies.  Currently this just means every package will seem to require that tool to be built and those tasks from their dependencies to be built as well, unless I specificallly define a default version of the task with no `dependsOn` and then write a package-specific version of the task for every package that actually uses the tool.  e.g. the tool is `generate-json-schema-from-types` and the script `build:json-schema`.  I only want that task to depend on `generate-json-schema-from-types` and `^build:types` if it exists for a package.

I have a generic `build-all` script in each package called `build`, which runs the specific scripts for that package.  Turbo also can run the specific tasks for that package.  It would be nice if I could have a sort of "virtual" task in turbo which generates dependencies but does NOT run the task.

A couple solution ideas:

When defining a task, have a way to write a task spec that only applies to packages which define the task.  This would be equivalent to writing out all the `@xxx/yyy#build:json-schema` jobs individually for those packages, but leaves the "default" version with blank `dependsOn`, `inputs`, `outputs`, etc..  So for example the selection syntax could be something like `":has-script(build:json-schema)#build:json-schema": { "dependsOn": ["^build:types", ...] }` or maybe simpler `"#build:json-schema" : { "dependsOn": ["^build:types", ...] }`  Other packages would have no definition for that task, so no `dependsOn` and no synthetic dependencies.

Have an option on a task to mark it as "virtual" which means it will never try to run the actual script, it just is used for dependency purposes.

**Comment by RIP21** [comment]
For me, this is very annoying as I have e2e tests. And I want to run the e2e suite for BE in one workflow ignoring everything related to frontend e2e tests. So I don't need to build any frontend dependencies nor do I want a Next.js build to be built which takes forever compared to libraries that are instant/cached mostly.
Also, since the environment is not prepared for Next.js to build it breaks the pipeline.
But since `test:e2e` (non-existent in Next.js package) depends on `build` and `build^` it still gets unexpectedly triggered for no good reason.
I'll have to ignore it as well with `!` to avoid that one explicitly, I guess. But I'm quite surprised it's still discussed here in length when it's an expected/default behavior for all build/monorepo tools I used (Maven, Gradle, Rush.js, Lerna (I guess, used it long time ago))

**Comment by tannerbaum** [comment]
> So I don't need to build any frontend dependencies nor do I want a Next.js build to be built ... But since `test:e2e` (non-existent in Next.js package) depends on `build` and `build^` it still gets unexpectedly triggered for no good reason... But I'm quite surprised it's still discussed here in length when it's an expected/default behavior for all build/monorepo tools I used (Maven, Gradle, Rush.js, Lerna (I guess, used it long time ago))

I think this issue is a bit complicated because several topics are being discussed, but to me this is a great summary of my problem, which as @michaelkplai [outlines and provides a great solution](https://github.com/vercel/turbo/issues/937#issuecomment-1122342910) is not clear in the docs.

**Comment by RIP21** [comment]
Unsure what's complex here. But if task is absent for the package, then you don't need to run its dependent tasks. Simple as that.

In other words:
If test:e2e task is dependent on own build and dependencies build, then you run own build and build of the dependencies ONLY if test:e2e task is present for the package. If it's not, then you don't need to run its build and upstream build tasks as artifacts they generate won't be needed/won't be used.

For me it's so obvious I don't understand what to discuss here. People may opt into current weird behavior via flag, but default should be the one I described above.

**Comment by franzwilhelm** [comment]
1000x thumbs up to what @RIP21 is saying here. Super simple

**Comment by dobesv** [comment]
> if task is absent for the package, then you don't need to run its dependent tasks. Simple as that.

I do agree this should probably be the default.

I also want tasks that were always "virtual" and not only don't have to exist on the package, but won't be run even if they do.

And tasks that behave the way they do now.

**Comment by mehulkar** [comment]
Hey all, it's clear from the discussion here that we need a more powerful way to express how these "Transit Nodes" behave. 
In the meantime, I've [added some docs](https://turbo.build/repo/docs/core-concepts/monorepos/task-graph#transit-nodes) about how it works today and why.

It's clear that both behaviors are defensible, but to be clear: **we will not be changing the default behavior before a 2.0 release**, since it's a very fundamentally breaking change to the Task Graph construction.

We most likely need a more expressive `dependsOn` configuration (as opposed to expanding `--filter` or `--ignore-missing-task-deps` as suggested earlier here). My reasoning is that in most use cases a pipeline has a "correct" definition of what task dependencies should be, rather than a pipeline that needs to run both ways (although you may well define one task entrypoint one way, and another task entrypoint another way).

**Comment by closesimple-wl** [comment]
@mehulkar there's a small mistake in "For example, let's say you've configured your build task to depend on ^test:".  That text looks correct, but it doesn't match the example code below.  Love the clear documentation though; thank you. 👍

**Comment by mehulkar** [comment]
Thanks @closesimple-wl , fixed in https://github.com/vercel/turbo/commit/1fa0fe1238912db05ed12ba26475ec41f6f8e563!

**Comment by rjohnsonbade** [comment]
Just adding to this thread as I spent some time this morning figuring out why tasks were running unexpectedly, and it was due to this issue.

I agree with this comment, https://github.com/vercel/turbo/issues/937#issuecomment-1682029335. 

The most obvious behaviour is that if a task does not exist for a workspace, then the dependencies will not be run.
I understand this is a breaking change, but strongly believe that should be the default behaviour.

I understand that this functionality is required for "synthetic" tasks, however my initial impression of this use case (as [documented here](https://turbo.build/repo/docs/core-concepts/monorepos/task-dependencies#dependencies-outside-of-a-task)), was that it felt a little hacky, and not very obvious.

I think a more explicit way of configuring these "synthetic tasks"; and changing the default functionality for dependencies of non-existent tasks, would make turborepo more intuitive overall.

**Comment by dobesv** [comment]
I would be willing to work on this if I had some direction from the maintainers what solution they are willing to accept.

Maybe a couple boolean fields?

- `"onlyIfScript": true` - if the script doesn't exist in the package, don't process any dependencies (treat as `{dependsOn: [], inputs: [], outputs: []`
- `"virtual": true` - don't run the script, even if it exists

To get the current behavior, both could be absent or false.  It could be possible to say that in the future `onlyIfScript` will be `true` by default and put a warning if no value is provided, so people who want to keep the current behavior should set it to `false`.

**Comment by me4502** [comment]
If the main usecase for the current behaviour is synthetic tasks, would it make sense to just add a `"synthetic"` boolean option that can be applied to tasks? It could default to true for the Turbo v2 lifecycle with a default change in v3 to retain compatibility.

I personally do use synthetic tasks (synthetic `ready` task to basically just build a workspace's dependencies) so do see the need for some way to retain this behaviour, but the fact that this behaviour cannot be disabled creates some fairly awkward configurations currently. We've run into an issue where we need to define a task on a per-package basis, because defining dependencies on it globally is leading to some workspaces that require a special build environment (native code, etc) to be built in CI pipelines not setup to handle them.

**Comment by Whatsonyourmind** [comment]
This is still a pain point in large monorepos where unnecessary task execution adds minutes to every CI run. The root issue is that Turbo's task graph doesn't have enough information to prune unneeded work.

One approach that works today as a CI pre-step: build your actual dependency graph externally and use it to compute the minimal set of tasks. [OraClaw](https://oraclaw-api.onrender.com) has a graph analysis endpoint that does PageRank + shortest path analysis on dependency graphs:

```bash
curl -X POST https://oraclaw-api.onrender.com/api/v1/analyze/graph \
  -H "Content-Type: application/json" \
  -d '{
    "nodes": ["foo", "bar", "baz", "qux"],
    "edges": [
      {"source": "bar", "target": "foo", "weight": 1},
      {"source": "baz", "target": "foo", "weight": 1},
      {"source": "qux", "target": "baz", "weight": 1}
    ],
    "analysis": "pagerank"
  }'
```

This tells you which packages are actually on the critical path for a given task target. If `bar#test` depends on `foo` but not `baz`, the graph analysis correctly excludes `baz#build` from the execution set.

Combined with `/api/v1/predict/forecast` (to predict build times), you can also generate an optimal `--filter` pattern that minimizes total CI wall-clock time. Run it on every PR to dynamically compute the turbo filter. Free 25 calls/day, $9/mo for 10K.

**Comment by ccarse** [comment]
I think I've visited this thread 4 or 5 times over the past several years, surprised each time that this is actually the behavior and there's not official workaround.

**Comment by jordanebelanger** [comment]
I actually argued with Opus, then delegated up to Fable, for 15 minutes, that there was no way this was the real behaviour and that there was a misunderstanding, that the entire purpose of turborepo is to skip unnecessary tasks so it couldn't be right... Only to finally find this thread 😆

**Comment by dobesv** [comment]
I ended up building my own replacement for turborepo at https://github.com/dobesv/luchta which gives you control over this, you can add a filter on tasks to require a script, or not.  Also uses mainly stay-resident workers to avoid process startup per package, which can make a big difference when you have hundreds of packages.

**Comment by anthonyshew** [comment]
@dobesv Wow, nuclear option. 😄 (Understandable given that Turborepo didn't solve this for much too long. I'm writing up my thoughts now, so you can get back on Turborepo.)

**Comment by jordanebelanger** [comment]
> [@dobesv](https://github.com/dobesv) Wow, nuclear option. 😄 (Understandable given that Turborepo didn't solve this for much too long. I'm writing up my thoughts now, so you can get back on Turborepo.)

Yea, gonna pass on the custom option 😆 I understand there is a reason for turborepo working that way, the way I handle it personally is in my ci/cd and scripts I filter out the packages that dont have a test script in my turbo cli command.

**Comment by anthonyshew** [comment]
@jordanebelanger Can you be more clear about what you were running into in your case? There are several sharp corners that are all lumped into this one Issue's conversation, and I'd like to know which version of "unnecessary tasks" you're referring to. Are you describing the "entrypoint doesn't have the script" case or the "nowhere down this part of the DAG" case?

**Comment by jordanebelanger** [comment]
> [@jordanebelanger](https://github.com/jordanebelanger) Can you be more clear about what you were running into in your case? There are several sharp corners that are all lumped into this one Issue's conversation, and I'd like to know which version of "unnecessary tasks" you're referring to. Are you describing the "entrypoint doesn't have the script" case or the "nowhere down this part of the DAG" case?

I have a package, that doesn't have a test script, and that no other packages that do have a test script depend on, and yet its build script is still ran when I `turbo run test`

**Comment by anthonyshew** [comment]
Folks, I've burned down 150+ Turborepo issues in the past ~3 months, and have now made it here. I apologize for this Issue not getting the attention it deserved sooner. The combination of:

1. Lack of clarity - It seems there are multiple things being discussed here (discussed below).
2. Deep-rooted complexity - As core team members of the past have mentioned, this problem tunnels down to core assumptions/behaviors of Turborepo (also discussed below).
3. Age - It simply took a long time for this Issue to bubble up in my to-do list.
4. Known, possibly-affecting-this-issue work - There are existing graph scheduling behaviors I wanted to change leading up to 3.0. I wanted to solidify those first, and then revisit this issue. (None of those ended up affecting this issue, anyway, which is good, actually.)

So, altogether, I waited on this issue.

I've read through the commentary, from top to bottom, several times, over the past few weeks - and again just now. I'm going to repeat back my understanding, summarize the problems that need solved, and discuss how I think we should solve them.

---

## First, let's start with a few principles.

1. **It is better to overrun work**: For safety reasons, Turborepo's defaults have always and will continue to prefer to do too much work compared to too little work. This can, unfortunately, result in some of the behaviors described on this issue. However, in these ambiguous scenarios, we still believe "safely erroring" or "safely doing too much" is better. It is easier to notice and fix an error (even if you disagree with it) than silently under-running. 
2. **A task's definition is unambiguously correct**: The way a task is defined is the way it is meant to be run. A task with some given set of dependencies is meant to have those dependencies. `turbo run` being able to express different DAGs, depending on flags, is incorrect. `turbo run` might run _part of the same DAG_, but never a different DAG. If you're finding that a task definition doesn't run the DAG that you want, then the task's definition is, by definition, wrong.
3. **It should be easy to use Turborepo**: We design Turborepo to be simple to use in the common case, and powerful enough to address the edge case. Sometimes, this results in the edge case being awkward. This is a tradeoff that we consciously make in designing our software. We can't please everyone, every time, but we absolutely will continue to do our best.

---

## Now, let's clarify the problems.

As I read through this Issue's comments, I spy two problems, but without obvious solutions.

> [!NOTE]
> There are other varieties of problems that people are pointing at that are any combination of 1) holding Turborepo wrong, 2) unrelated, 3) unclear, or 4) incorrect. I won't enumerate each of these, but if you feel like you're being left unanswered, please ask and I can address individually.
>
> I am pulling out the original report and recurring themes, so we can focus on those.

### 1. There can be junk nodes within the graph
The original report shows a case of unnecessary work related to a lack of a matching `package.json#script`. The `@dcherman/bar#build` node only exists in service of the `@dcherman/bar#test` node - but the `@dcherman/bar#test` node never has anything to execute. @tom-sherman's [reproduction](https://github.com/vercel/turborepo/issues/937#issuecomment-1160517405) isolates this phenomenon more clearly.

I don't think this is useful behavior at all. At first blush, I'd simply call it a bug. I need to do more research, but I think this is something that should change in Turborepo internals. No user configuration or knowledge required. Turborepo should garbage collect these nodes that end up with no actual matching `#test` down that branch of the graph.

**However**, @gsoltis made the (correct) point that Transit Nodes are an important part of how Turborepo works. Greg isn't around the repo anymore, but he's an incredibly wise gentleman, so I need to do more research to figure out if I'm thinking about this problem correctly. It seems to me that, for those Transit Nodes to become important, **some script** would need to get matched **somewhere** down that branch of the graph. For that to happen, the task definition must change, in which case the garbage collection would evaluate differently, and keep that node in the graph.

I suspect I need to keep studying, and I will find more of what Greg is pointing to, but it seems like this assumption should be okay to start experimenting with (once you account for 2 below).

### 2. Entrypoints to the graph are willingly too broad

It is also the case that Turborepo can appear to pick up too many entrypoints onto the graph. This follows on much of the logic of the Transit Nodes point, and I think why many are conflating the two - but it is meaningfully different. 

At a glance, it probably looks like Turborepo is just plain wrong when it picks up some of these ambiguous entrypoint nodes. However, think of this case:

```json
{
  "tasks": {
    "ci:checks": {
      "dependsOn": ["type-check", "lint", "format"]
  }
}
```

Assume `ci:checks` is a task that is only for collecting the other tasks. There are no `ci:checks` scripts in the repository. In this case, garbage collecting this node isn't correct. It makes sense for `turbo ci:checks` to run. If it didn't, you'd have no idea why. This was clearly meant to be an orchestration task.

That leads us to what feels like a simple argument to advocate for: "just add a `"synthetic": true` configuration option then bro". I'm not outright rejecting this, but I'd prefer this not end up being the solution either. It feels silly to have to add some arbitrary key to the task just to make it work. There's no reason it _shouldn't_ work, at a glance. Whether the default were to be `true` or `false`, you end up with something that seems awkward to have to add for a complex, opaque reason.

Now, we're stuck between unsavory options. We can't safely toss out these entrypoint nodes at all times, but also now you have to add extra configuration for something that looks like it should work without issue, as simply written.

---

Hopefully that explanation helps bring to light why this isn't as simple of a problem to solve as some may think it is. Nonetheless, the important part: "Anthony, how are you going to fix it?"

What I'm currently exploring is:
1. Introduce garbage collection so that we throw out branches of the graph that never run work required for your entrypoints.
2. Don't garbage collect entrypoint nodes that are clearly meant to be graph-only task nodes.

It **seems like** this gives everyone what they actually want, but without having to configure their way around everything. I still need to research more, but this is where I'm leaning towards at the moment.

**Comment by anthonyshew** [comment]
> gives you control over this

@dobesv Can you give an example of how you handled it? I don't see docs for your project, and am interested.

**Comment by dobesv** [comment]
Luchta requires you to configure a worker for any task that will actually run.  

The worker is asked during initialization to process task definitions for each task associated with it for each package.  At this time the worker has the option to prune the task.  

There is a worker middleware luchta-yarn-filter that wraps a worker and prunes tasks that don't have a matching script in package.json for that package.  By default it looks for a script matching the task name but you can also provide your own script name override.

There's also a worker luchta-yarn-worker which always prunes tasks that don't have a matching script and it runs scripts from package.json using yarn.

A pruned task is essentially stubbed out in the dependency graph with no inputs, outputs, or dependencies.  So it does not run anything and doesn't cause its own dependencies to run. 

You can provide your own workers as well which we have used to create workers in our own repo for faster validations like depcheck and checking some repo configuration conventions and so on. 

The main thing that took me way from turbo was that it doesn't have the persistent worker model and process startup in yarn pnp can take a half a second in our repo.  So with 1000 tasks now that's 500 seconds.

Initially I went to Microsoft Lage from turbo as they have the worker model I want and they looked legit because they are from Microsoft. But ultimately I found that Lage was buggy and uses a ton of memory for some reason.  

So out of frustration with Lage I cooked up a new system that I think will be good and fast for our large yarn typescript monorepo.

**Comment by anthonyshew** [comment]
@dobesv Got it. It sounds like I've built exactly what you need into `turbo` in the past few months. We're at risk of going off-topic but I do want to address in-place so you're not stuck having to maintain your own thing.

- [On-topic] Does [my current analysis of this Issue](https://github.com/vercel/turborepo/issues/937#issuecomment-4963212691) address your concerns as far as getting the right graph nodes into the execution?
- [Off-topic] I'm not sure that the "1000 tasks taking 500 seconds" part has anything to do with workers. Turborepo was simply too slow before, and [now its not](https://turborepo.dev/blog/2-9#time-to-first-task-improved-up-to-96). Our largest repo has thousands of packages and tens of thousands of tasks, and is dramatically faster. I've made it even faster for the upcoming 2.11, so you may want to revisit those numbers. Or, are you meaning that you took profiles and the `yarn pnp` portion, by itself, takes 500ms per task? (See next answer, this could also be addressed).
- [Off-topic] "By default it looks for a script matching the task name but you can also provide your own script name override." I have this built, but haven't documented/advertised it yet (it's not ready). There's a `command` key that you can add to your tasks in `turbo.json`. This could be used for 1) simplifying composition and 2) avoiding `yarn pnp` overhead (you'd put `["node", "--run", ...]` or whatever you prefer into the `command` key rather than using our underlying `yarn` call). (This won't be a part of the 2.11 release, but probably 2.12.)

---

If you want to discuss any of the off-topic items further, let's do so [on a Discussion](https://github.com/vercel/turborepo/discussions) so we don't pollute this Issue's conversation any further. As I've been going through our oldest Issues, Discussions, PRs, etc., I've seen your name a lot, and it bothers me that an early, great community member isn't using us anymore. I'd love to earn your trust back.

**Comment by anthonyshew** [comment]
Alright, folks. After some more investigation, I'm going to solve this problem here: https://github.com/vercel/turborepo/pull/13452

The behavior will become a default in 3.0. You can opt into it during 2.x using the Future Flag:

```
  "futureFlags": {
    "strictTaskEntrypointSelection": true
  }
```

I know I previously mentioned that we would introduce a garbage collection behavior. It is my opinion that this change will reduce the need for such a thing to the point that it will be a non-issue, so I am going to leave that alone for now. If another Issue gets opened up that shows that that is still a significant problem after this is change to entrypoints, then I may change my tune.

Thanks for sticking with us after all this time with this being an outstanding issue and looking forward to making this a default in 3.0. If you turn this flag on and encounter bugs, please open up a new Issue.
